README.md

```diff
@@ -1,2 +1,1088 @@
+set of goals and an initial algorithm (usually the `idle_robot`).
+These parameters are validated to make sure the maze is solvable and the
+starting point is valid (start position is connected to the goal, robot doesn't
+start facing a wall, the goal is not an empty set of cells).
+
+The `Simulator` holds 2 [`ExtendedMaze`s](#simmaze) - 1 is the full maze, and 1
+is the maze as seen by the robot.
+
+The `Simulator` provides the following functionality:
+
++ `restart(alg: Algorithm)` - Restart the simulation with the provided
+  algorithm. This resets most internal counters and the robot's maze and
+  consumes the robot's `READY` action.
+
+  If the robot has no actions, the robot's first action is not `READY`, or the
+  robot encounters an error before te first action, the simulation status
+  becomes `ERROR` and a `RuntimeError` is raised.
+
+  Otherwise, the simulation's status after this function is `READY`.
++ `step()` - Perform a single action from the robot.
+
+  If the simulation's status is `ERROR`, no action is performed.
+
+  The action is retrieved by calling `robot.send(robot_state)` where
+  `robot_state` is a [`RobotState`](#simrobots) instance representing the
+  current robot's position and heading.
+
+  After this function, the simulation's status is:
+
+  + `ERROR` if the robot encounters an error or performs an illegal action (for
+    example, attempts to go through a wall), in this case, a `RuntimeError` is
+    also raised.
+  + `FINISHED` if the robot raised `StopIteration` and is located in one of the
+    goal cells.
+  + `IN_PROGRESS_FOUND_DEST` if the robot found the goal but has more actions.
+  + `IN_PROGRESS` otherwise.
+
++ `maze` - The simulation's maze.
+
+  The optimal route, calculated by Dijkstra's algorithm with full knowledge of
+  the maze is saved in this maze's `route` property.
++ `begin` - A `(row, col, heading)` tuple of the robot's initial state.
++ `end` - A set of `(row, col)` coordinates of the goal cells. This set is
+  passed to the robot.
++ `robot_maze` - The maze, as seen by the robot with all additional information
+  added by the robot.
++ `robot_pos` - The robot's current position + heading.
++ `status` - The current simulation status.
++ `counter` - A set of counters maintained by the simulator. These counters
+  include visited cell count, action weight, and action count for both the
+  current run (since the last `RESET` action) and the total for entire
+  simulation (since the lase `simulator.reset(...)` call).
++ `connected(...)` - Check if cells are connected.
+
+### `sim.front`
+
+Defines classes and utility functions for writing frontends with this package.
+
+The [command-line types](#command-line-types) and the base classes for the
+supported custom entrypoints are defined here.
+
+### `sim.tools`
+
+The [provided tools](#provided-tools) are implemented here.
+
+### `sim.unionfind`
+
+A simple Union-Find (Disjoint Sets) implementation.
```

Comment:
I think the data structure is called a Disjoint Set (not sets).

Reply:
Also with a hyphen :/